unblob/handlers/archive/arj.py

```diff
@@ -0,0 +1,184 @@
+
+        file.seek(start_offset)
+        main_header = cparser.arj_header(file)
+        file.seek(start_offset + main_header.first_hdr_size + len(cparser.basic_header))
+
+        # Read main ARJ file name, comment, and CRC
+        _ = read_until_null(file)
+        _ = read_until_null(file)
+        _ = file.read(4)
+
+        # TODO: Handle the extended header
+        extended_header = cparser.extended_header(file)
+        if extended_header.size != 0:
+            raise ARJExtendedHeaderException
+
+        return file.tell()
+
+    def _read_arj_files(
+        self, cparser: cstruct, file: io.BufferedIOBase, start_of_files: int
+    ) -> int:
+        while True:
+            # Save where we're starting
+            start = file.tell()
+            basic_header = cparser.basic_header(file)
+
+            if basic_header.size == 0:
+                # We've reached the final empty file header. This is where we want to be.
+                return file.tell()
+
+            # Cursor back to the start of the files section.
+            file.seek(start)
+            file_header = cparser.file_header(file)
+
+            # Use the first_hdr_size to read past the header to the filename and comment strings.
+            file.seek(start + len(cparser.basic_header) + file_header.first_hdr_size)
+
+            # Read the filename, comment, and CRC
+            _ = read_until_null(file)
+            _ = read_until_null(file)
+            _ = file.read(4)
+
+            # TODO: Handle cases where there is an extended header.
+            extended_header = cparser.extended_header(file)
+            if extended_header.size != 0:
+                raise ARJExtendedHeaderException
+
+            # Read past the file contents
+            _ = file.read(file_header.compressed_size)
+
+    def calculate_chunk(
+        self, file: io.BufferedIOBase, start_offset: int
+    ) -> Union[ValidChunk, UnknownChunk]:
+
+        # Not using the inbuild self.parse_header() here, because we need access to a few different
+        # structures at various points.
+        cparser = cstruct()
+        cparser.load(self.C_STRUCTURES)
+
+        try:
+            # Read past the main header.
+            end_of_main_header = self._read_arj_main_header(cparser, file, start_offset)
+            end_of_arj = self._read_arj_files(cparser, file, end_of_main_header)
+        except ARJFileLengthException:
+            return UnknownChunk(
+                start_offset=start_offset, reason="This is a zero-sized ARJ."
+            )
+        except ARJExtendedHeaderException:
+            return UnknownChunk(
+                start_offset=start_offset,
+                reason="Main ARJ header contains extended_header, which we don't currently handle.",
+            )
+        except EOFError:
+            return UnknownChunk(
+                start_offset=start_offset,
+                reason="File ends before ARJ file resolves.",
+            )
```

Comment:
All of this can be replaced by warning messages now that we are no longer emitting UnknownChunk structures.